rfcs/20200712-tfrt-kernel-fallback.md

```diff
@@ -0,0 +1,647 @@
+
+```
+%tft_c = "tfd.kernel_fallback"(%tft_a, %tft_b) {
+    _op_name = "Conv3D",
+    attr1_name="data_format", attr1_value="string$NDHWC",
+    attr2_name="strides", attr2_value="list(i32)$1,1,1,1,1",
+    attr3_name="dilations", attr3_value="list(i32)$1,1,1,1,1",
+    attr4_name="padding", attr4_value="padding$SAME"}: (!tfd.tensor, !tfd.tensor) -> !tfd.tensor
+```
+
+For example, `dilations` attribute here has a value of `[1, 1, 1, 1, 1]`.
+Note: TFRT orders attributes by name, alphabetically, which is why we use `attrN_value` and `attrN_name` pattern pair.
+
+## Reusing Kernels
+
+TensorFlow currently reuses kernels instantiated for a particular node in a
+graph. It would be nice to have this optimization for Kernel fallback as well.
+
+BEF executor keeps track of offsets within a BEF file. We can use this offset to
+cache corresponding kernel objects.
+
+We should make sure that Kernel Fallback is thread safe when reusing kernel
+objects since Compute for the same kernel can be called from multiple threads.
+We can take a simple approach and support kernel cache only for stateless
+kernels. Stateless kernels only update `OpKernelContext` and not `OpKernel`
+state itself.
+
+## C API Integration
+
+Modular TensorFlow effort aims to break up giant monolithic TensorFlow binaries
+into smaller shared libraries. Specifically, James (@sjamesr) and Gunhan
+(@gunhan) looked at splitting out kernels out of TensorFlow core. Initial Kernel
+C API definition is at
+[kernel.h](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/c/kernels.h)
+and its implementation is at
+[kernel.cc](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/c/kernels.cc?q=kernels.cc).
+
+Kernel Fallback should support kernels migrated to C API as well. We can
+implement this support behind the C API, so that we don’t have to update
+individual kernels.
+
+### C API multiple implementation structure
+
+There are a few important takeaways from current kernel C API implementation
+that will impact decisions in the document:
+
+1.  We register a
+    [COpKernel](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/c/kernels.cc;l=104?q=copkernel)
+    object (with TensorFlow op kernel registry) for _any_ kernel defined using
+    the C API.
+1.  `OpKernelContext` and `OpKernelConstruction` are passed around as opaque
+    pointers on the C API surface (they get cast to `TF_OpKernelContext` and
+    `TF_OpKernelConstruction` aliases).
+1.  Most of the functions just provide accessors into
+    `OpKernelContext`/`OpKernelConstruction` types.
+
+Given current API structure, we can consider two approaches going forward:
+
+1.  TFRT fully supports all functionality available in the C API. This way any
+    kernel defined using the C API would be automatically available using either
+    full TensorFlow or the TFRT kernel fallback.
+1.  Certain functionality is only available with TF backend. TFRT C API
```

Comment:
how would this "fall back to full TF..." work for the mobile case?

Reply:
My main point being that the second option that relies on current TF runtime as fallback is likely not feasible given the goals of kernel fallback for mobile, and/or would require the runtime fallback work to target mobile as well (which is an expressed non-goal at this point).

Because of this, and general cleanliness, I agree with your suggestion below that option 1 should be preferred.

Reply:
This paragraph is talking about a choice that can be made at compile time:

* For mobile, we can build with kernel fallback.
* For non-mobile usecase, we can build with full TF.

In more detail: C API will have 2 implementations selected at compile time. Therefore, every kernel implemented using the C API will automatically be supported with kernel fallback (first bullet point above) or full TensorFlow (second bullet point).